python/ray/tune/suggest/__init__.py

```diff
@@ -93,7 +94,15 @@ def _import_sigopt_search():
             f"Got: {search_alg}")
 
     SearcherClass = SEARCH_ALG_IMPORT[search_alg]()
-    return SearcherClass(**kwargs)
+
+    search_alg_defined_args = inspect.getfullargspec(SearcherClass).args
```

Comment:
from [the docs](https://docs.python.org/3/library/inspect.html#inspect.getfullargspec):
> Note that `signature()` and Signature Object provide the recommended API for callable introspection, and support additional behaviours (like positional-only arguments) that are sometimes encountered in extension module APIs. This function [`getfullargspec`] is retained primarily for use in code that needs to maintain compatibility with the Python 2 inspect module API.

@kaushikb11 what you've written in this PR should work but perhaps we can do something like this instead?
```python
from inspect import signature, Parameter


def get_function_kwargs(callable):
    # untested
    all_parameters = signature(callable).parameters
    kwargs = [param for param in all_parameters if param.kind == Parameter.VAR_KEYWORD]
    return kwargs
```

Reply:
@sumanthratna Thanks! 
> kwargs = [param for param in all_parameters if param.kind == Parameter.VAR_KEYWORD]

I was thinking why is doing `if param.kind == Parameter.VAR_KEYWORD` required? Could we just do
`kwargs = [param for param in all_parameters]`. 
Also, tried implementing `if param.kind == Parameter.VAR_KEYWORD`, it returns an empty array.
This works,
`kwargs = [param.name for param in all_parameters.values() if param.kind == Parameter.POSITIONAL_OR_KEYWORD]`

Will be adding this function in `ray.utils`.

Reply:
> Could we just do `kwargs = [param for param in all_parameters]`

I'm not sure why `VAR_KEYWORD ` returned an empty list, but you're right, I don't think we need to filter by `Parameter.kind`. I think something like this might work:
```python
# in the utils file:
from inspect import signature

def get_function_args(callable):  # note the name change!
    # untested
    all_parameters = frozenset(signature(callable).parameters)
    return list(all_parameters)

# in the __init__ file:
def create_searcher(search_alg, **kwargs):
    # ...
    search_alg_args = get_function_args(SearcherClass.__init__)
    trimmed_kwargs = {k: v for k, v in kwargs.items() if k in search_alg_args}
    return SearcherClass(**trimmed_kwargs)
```

Reply:
@sumanthratna Noted! Updated for `tune.create_scheduler` as well. I was wondering if we could add some kind of validation of params for the search algorithms. For example, `grid_search` is not supported for `AxSearch`, `BayesOpt` doesn't support `Integer` data type. It raises an error during the `tune.run`. If there could be a way to get functionalities that are not supported for `search_alg` for validation before running `tune.run`.

cc: @richardliaw

Reply:
@kaushikb11 that's a really good idea and also very useful—I'm not sure about the best way to implement this, but this is what I'm thinking for now:
1. create a function decorator that inspects the source and locates all Exceptions/Errors that are raised
2. creates a flow control graph of the function to be executed
3. quickly trace this graph with the given inputs and determine if any errors would be raised (sort of like TensorFlow 2's autograph/JIT tracing)
    - if no errors are raised, yay! proceed
    - if errors are raised, determine which variable triggered the exception and determine allowed values and update the kwarg to an allowed value
    - I think we _need_ steps 1 and 2 because if we trace without simplifying the code graph, we'll end up actually executing code during the trace (which would mean code is executed twice, which we don't want)

@richardliaw thoughts?

Reply:
on second thought, I'm not sure we want to do this (maybe I'm just over-complicating :)):

suppose we have a search alg that doesn't allow `Integer` domains when `mode='min'` (purely hypothetical). What should we do in this case? Should we change `Integer` to another domain and/or should we change `mode` to `'max'`? Both sound intrusive and could confuse users a lot. I think the only way that would save users from confusion would be to not actually patch the user-provided kwargs, and instead raise an error in the shim function or something similar. However, I don't see the benefit in this: now we're just raising an error in the shim instead of the actual `Class.__init__`

@kaushikb11 am I misunderstanding your suggestion?

Reply:
@sumanthratna:
Here's my use case: https://github.com/uber/ludwig/pull/1001
We are defining the parameters in the `yaml` file for the `RayTuneExecutor` for `search_algo`, `sampling`, etc. I wish to check if all the parameters are supported by the `search_alg` before the `tune.run` execution, according to that, we could warn/suggest Users if we are skipping sampling those parameters or raise an Error before the execution.

Reply:
Let's leave the validation of search params for another PR, and avoid expanding the scope of this one.

Reply:
@richardliaw: Yup, will create a separate Issue on this in some time.